docs(zeebe): add docs for dynamic scaling api (#2861)
camunda/camunda-docs: docs/self-managed/zeebe-deployment/operations/cluster-scaling.md

---
id: cluster-scaling
title: "Cluster Scaling"
description: "How to scale a Zeebe cluster"
---

> :::note

Reviewer: :wrench: 

```suggestion
:::warning
```

This is an experimental feature.
:::

> Zeebe allows scaling an existing cluster. When new brokers are added, the partitions will be redistributed to the new brokers thus distributing some load. The cluster can also be scaled down to remove extra brokers.

Reviewer: ```suggestion
Zeebe allows scaling an existing cluster by adding or removing brokers. Partitions are automatically redistributed over the new set of brokers to spread the load evenly.
```

Zeebe provides a REST API to manage the cluster scaling. The cluster management API is a custom endpoint available via [Spring Boot Actuator](https://docs.spring.io/spring-boot/docs/3.1.x/reference/htmlsingle/#actuator.endpoints). This is accessible via the management port of the gateway.

## Configuration

> - This is an experimental feature. To use the feature set `ZEEBE_BROKER_EXPERIMENTAL_FEATURES_ENABLEDYNAMICCLUSTERTOPOLOGY` to `true`.

Reviewer: :thought_balloon: I think we should explain what happens when users disable the feature after changing the cluster topology.

Reviewer: If we want to go the extra mile, we could also explain how to do "rollbacks" by deleting the topology files.

Author: > If we want to go the extra mile, we could also explain how to do "rollbacks" by deleting the topology files.

This would be risky. Let's add it in the next iteration.

Author: > Also, when this is enabled, fixed partition distribution is not respected!

:+1: Good point. I will add that.

## Scaling API

To scale up, start new brokers and use the scaling api to add the new brokers to the cluster and distribute data to them.

### Request

```
POST actuator/cluster/brokers/
[
  <brokerId1>, <brokerId2>,..
]
```

> The input is a list of broker Ids that will be in the final cluster after scaling.

Reviewer: ```suggestion
The input is a list of *all* broker ids that will be in the final cluster after scaling.
```

<details>
  <summary>Example request</summary>

```
curl --request POST 'http://localhost:9600/actuator/cluster/brokers' \
-H 'Content-Type: application/json' \
-d '["0", "1", "2", "3"]'
```

</details>

### Response

The response is a json object

```
{
  changeId: <changeId>
  currentTopology: [...]
  plannedChanges: [...]
  expectedTopology: [...]
}
```

> - `changeId`: Id of the changes initiated to scale the cluster. This can be used to monitor the progress of scaling operation.

Reviewer: :thought_balloon: Should we document that this is an increasing counter?

---
id: cluster-scaling
title: "Cluster scaling"
description: "Scale an existing cluster by adding or removing brokers."
---

> :::warning

Reviewer 2: Is there anything else we should be noting here? Is this an alpha feature? If this is experimental, could it cause any other issues? Will I lose any data by trying this?

Author: This will be in `8.4.0-alpha1`. Since this is an experimental feature, we do not recommend to use it in production. Should we add this? Is there a standard disclaimer for such features?

Reviewer 2: I added it to the note 👍

Author: Thank you @christinaausley
